File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const genreTotal = {};
    const genreSongs = {};
    
    // 장르별 총합과 노래 목록 계산
    for (let i = 0; i < genres.length; i++) {
        const g = genres[i];
        genreTotal[g] = (genreTotal[g] || 0) + plays[i];
        if (!genreSongs[g]) genreSongs[g] = [];
        genreSongs[g].push([plays[i], i]);
    }
    
    // 장르 총합 내림차순 정렬
    const sortedGenres = Object.keys(genreTotal).sort((a, b) => genreTotal[b] - genreTotal[a]);
    
    const answer = [];
    for (let g of sortedGenres) {
        // 각 장르 노래 재생수 내림차순, 번호 오름차순 정렬 후 상위 2개
        genreSongs[g].sort((a, b) => a[0] === b[0] ? a[1] - b[1] : b[0] - a[0]);
        answer.push(genreSongs[g][0][1]);
        if (genreSongs[g].length > 1) answer.push(genreSongs[g][1][1]);
    }
    
    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        HashMap<String, Integer> genreTotal = new HashMap<>();
        HashMap<String, List<int[]>> genreSongs = new HashMap<>();
        
        // 장르별 총합과 노래 목록
        for (int i = 0; i < genres.length; i++) {
            String g = genres[i];
            genreTotal.put(g, genreTotal.getOrDefault(g, 0) + plays[i]);
            genreSongs.computeIfAbsent(g, k -> new ArrayList<>()).add(new int[]{plays[i], i});
        }
        
        // 장르 총합 내림차순 정렬
        List<String> sortedGenres = new ArrayList<>(genreTotal.keySet());
        sortedGenres.sort((a, b) -> genreTotal.get(b).compareTo(genreTotal.get(a)));
        
        List<Integer> answer = new ArrayList<>();
        for (String g : sortedGenres) {
            // 각 장르 내 정렬: 재생수 내림차순, 인덱스 오름차순
            List<int[]> songs = genreSongs.get(g);
            songs.sort((a, b) -> a[0] == b[0] ? a[1] - b[1] : b[0] - a[0]);
            answer.add(songs.get(0)[1]);
            if (songs.size() > 1) answer.add(songs.get(1)[1]);
        }
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}
